src/main.rs

```diff
@@ -66,99 +66,134 @@ async fn execute_traces(
             continue;
         }
 
-        info!("TRACING block {block_number} with Base. It has {tx_count} transactions");
-        let using_cached_trace = !redo_traces && base_trace_path(block_number).exists();
-        let base_trace_result = if !using_cached_trace {
-            base_juno.trace_block(block_number).await
-        } else {
-            Ok(read_base_trace(block_number).await)
-        };
-
-        // First branch is a succesful block trace with Base
-        // Second branch is an unexpected crash by Juno/Blockifier/nNative
-        match base_trace_result {
-            Ok(base_report) => {
-                if base_report.result != TraceResult::Success {
+        let base_trace_result =
+            trace_base(redo_traces, block_number, tx_count, &mut base_juno).await?;
+
+        info!("Switching from Base Juno to Native Juno");
+        base_juno.ensure_dead().await?;
+        let native_trace = trace_native(block_number, tx_count, &simulation_flags).await?;
+
+        if let Some(native_trace) = native_trace {
+            let comparison =
+                generate_block_comparison(block_number, base_trace_result, native_trace);
+            log_comparison_report(block_number, comparison).await;
+        }
+    }
+
+    Ok(())
+}
+
+async fn trace_base(
```

Comment:
Could you please add a small description as a comment or update the function name that this function will use caching